### DISCUSSION TITLE
How to upload a file which is then used by chatgpt to answer questions?

### DISCUSSION THREAD

**Discussion by Jerry0022** [SemlerPDX/OpenAI-VoiceAttack-Plugin#7] [Q&A] [answered]
Hey, 

I have a preformatted keybinding file. I managed to use File.Upload with "fine-tune". But when I ask Voiceattack / OpenAI Plugin what is the key for a certain command then it answers it can't access the file or gives me other random answers.
How can I make this work? 
When I use ChatGPT, just uploading a file in the chat, then it works for me. How can I transfer this use case to the OpenAI plugin?
Is fine-tune only work with the gpt specific format, or can I somehow use it also for such a "random" file content.
(the file is too large to send with each speach command as text and would produce a lot of overhead)

Basically I want to do what's shown here: https://www.youtube.com/watch?v=xo_iRQaFxb0

**Comment by SemlerPDX** [comment, marked as answer]
Please read more about what this context in my plugin is for here:
https://github.com/SemlerPDX/OpenAI-VoiceAttack-Plugin/wiki/Plugin-Context#file-request

Fine-tuning is an advanced function of OpenAI and a given model, and not exactly applicable to your use case scenario.  You could, however, create a custom system prompt and then provide a series of example inputs/responses which could be you stating a set of keybinds, and the response of, "Understood. I will note these keybinds for later" (for each input which contains a block of keybinds, however many you can make fit within the context limit of the model you are using).

Read more about user input and response examples here:
https://github.com/SemlerPDX/OpenAI-VoiceAttack-Plugin/wiki/All-VoiceAttack-Variables#openai_userinput_example
https://github.com/SemlerPDX/OpenAI-VoiceAttack-Plugin/wiki/All-VoiceAttack-Variables#openai_chatbotoutput_example

You can find an example of what a Fine-Tuning file should look like (in JSON format) in the example folder included with my plugin here:
`%ProgramFiles(x86)%\VoiceAttack\Apps\OpenAI_Plugin\tuning`
...or if you have the Steam version of VoiceAttack:
`%ProgramFiles(x86)%\Steam\steamapps\common\VoiceAttack\Apps\OpenAI_Plugin\tuning`

There are big things on the horizon regarding feeding files to OpenAI systems, but they are not here yet - they are still in beta, and require a fair bit of testing and development on my part in order to properly incorporate into a speech-to-text and text-to-speech system like I have so far with the other OpenAI systems.

If you have any follow up questions, please let me know and I'll be happy to help if I can.  It seems GitHub only offers me the ability to comment and close this thread, so please use the forum section of my plugin on GitHub to open a topic in a discussion format where we can chat back-and-forth if you want.

-Sem

**Comment by Jerry0022** [comment]
Thanks Sem, yeah, I looked into the capabilities of fine-tuning and intensively in your documentation, but couldn't find out if this use case can made to work, obviously not like this.
Appreciate your workaround, I will have a look. 
The other way would be I upload the keybinding file to the gpt assistant setup online, hopefully via an api, I will search later. Then my understanding is I could use the assistant parameter to access this one?

**Reply by SemlerPDX** [reply]
The Assistant API is still in Beta (iirc) and is not part of the OpenAI API Plugin for VoiceAttack (yet).  You can definitely try it out in the Playground and see if it is the right kind of tool for you, however if you wanted something integrated into my existing VoiceAttack systems, the workaround I described would more than likely suffice.

Best wishes and good luck!
